classes/security/authorization/internal/PublicationRequiredPolicy.php

```diff
@@ -16,25 +16,27 @@
 
 use APP\core\Application;
 use APP\facades\Repo;
-use APP\publication\Publication;
-
 use PKP\security\authorization\AuthorizationPolicy;
 use PKP\security\authorization\DataObjectRequiredPolicy;
 
 class PublicationRequiredPolicy extends DataObjectRequiredPolicy
 {
+    protected ?string $publicationParameterName;
+
     /**
      * Constructor
      *
      * @param PKPRequest $request
      * @param array $args request parameters
-     * @param string $publicationParameterName the request parameter we expect
-     *  the submission id in.
+     * @param ?string $publicationParameterName the request parameter we expect
+     *  the submission id in. Pass null to look for the current publication in an authorized
+     *   submission object instead.
      * @param null|mixed $operations
      */
-    public function __construct($request, &$args, $publicationParameterName = 'publicationId', $operations = null)
+    public function __construct($request, &$args, $publicationParameterName = null, $operations = null)
     {
-        parent::__construct($request, $args, $publicationParameterName, 'user.authorization.invalidPublication', $operations);
+        parent::__construct($request, $args, $publicationParameterName ?? '', 'user.authorization.invalidPublication', $operations);
```

Comment:
I can make this change, but it means my code will now impact 10 policies that extend `DataObjectRequiredPolicy`. Maybe there won't be a problem, but I never feel confident making such a change with the base security policy classes because it's very hard to test for side effects. Happy to do whatever you think is best here.

Reply:
I'd be confident about this (i.e. if it causes downstream problems I'll deal with it :smile:).

It tightens up error handling, establishes `null` as a valid `$parameterName` indicating "we dont do things that way, thanks".
```diff
diff --git a/lib/pkp/classes/security/authorization/DataObjectRequiredPolicy.php b/lib/pkp/classes/security/authorization/DataObjectRequiredPolicy.php
index be8c1e290d..2fca04d274 100644
--- a/lib/pkp/classes/security/authorization/DataObjectRequiredPolicy.php
+++ b/lib/pkp/classes/security/authorization/DataObjectRequiredPolicy.php
@@ -22,8 +22,7 @@ class DataObjectRequiredPolicy extends AuthorizationPolicy
     /** @var array */
     public $_args;
 
-    /** @var string */
-    public $_parameterName;
+    public ?string $_parameterName;
 
     /** @var array */
     public $_operations;
@@ -56,11 +55,11 @@ class DataObjectRequiredPolicy extends AuthorizationPolicy
      *
      * @param PKPRequest $request
      * @param array $args request parameters
-     * @param string $parameterName the request parameter we expect
+     * @param ?string $parameterName the request parameter we expect
      * @param string $message
      * @param array $operations Optional list of operations for which this check takes effect. If specified, operations outside this set will not be checked against this policy.
      */
-    public function __construct($request, &$args, $parameterName, $message = null, $operations = null)
+    public function __construct($request, &$args, ?string $parameterName, $message = null, $operations = null)
     {
         parent::__construct($message);
         $this->_request = $request;
@@ -114,7 +113,7 @@ class DataObjectRequiredPolicy extends AuthorizationPolicy
         $router = $this->_request->getRouter();
         switch (true) {
             case $router instanceof \PKP\core\PKPPageRouter:
-                if (ctype_digit((string) $this->_request->getUserVar($this->_parameterName))) {
+                if ($this->_parameterName !== null && ctype_digit((string) $this->_request->getUserVar($this->_parameterName))) {
                     // We may expect a object id in the user vars
                     return (int) $this->_request->getUserVar($this->_parameterName);
                 } elseif (!$lookOnlyByParameterName && isset($this->_args[0]) && ctype_digit((string) $this->_args[0])) {
@@ -125,19 +124,20 @@ class DataObjectRequiredPolicy extends AuthorizationPolicy
 
             case $router instanceof \PKP\core\PKPComponentRouter:
                 // We expect a named object id argument.
-                if (isset($this->_args[$this->_parameterName])
+                if ($this->_parameterName !== null && isset($this->_args[$this->_parameterName])
                         && ctype_digit((string) $this->_args[$this->_parameterName])) {
                     return (int) $this->_args[$this->_parameterName];
                 }
                 break;
 
             case $router instanceof \PKP\core\APIRouter:
-                $handler = $router->getHandler();
-                return $handler->getParameter($this->_parameterName);
+                if ($this->_parameterName !== null) {
+                    $handler = $router->getHandler();
+                    return $handler->getParameter($this->_parameterName);
+                }
                 break;
 
-            default:
-                assert(false);
+            default: throw new Exception('DataObjectRequiredPolicy does not support routers of type ' . get_class($router) . '!');
         }
 
         return false;
diff --git a/classes/security/authorization/OjsIssueRequiredPolicy.php b/classes/security/authorization/OjsIssueRequiredPolicy.php
index 2fa69432d5..50451eff42 100644
--- a/classes/security/authorization/OjsIssueRequiredPolicy.php
+++ b/classes/security/authorization/OjsIssueRequiredPolicy.php
@@ -92,7 +92,7 @@ class OjsIssueRequiredPolicy extends DataObjectRequiredPolicy
         $router = $this->_request->getRouter();
         switch (true) {
             case $router instanceof \PKP\core\PKPPageRouter:
-                if (ctype_digit((string) $this->_request->getUserVar($this->_parameterName))) {
+                if ($this->_parameterName !== null && ctype_digit((string) $this->_request->getUserVar($this->_parameterName))) {
                     // We may expect a object id in the user vars
                     return (int) $this->_request->getUserVar($this->_parameterName);
                 } elseif (isset($this->_args[0])) {
```

Reply:
Thanks for the diff!